### PR TITLE
GPU finite difference

### LINKED ISSUE #1171: GPU finite-difference operators
Two options

1. Oceananigans approach: launch a thread for every node. Conceptually this is the simplest, as each point can be evaluated independently.
   - Unlike Oceananigans we don't use halos for boundary conditions. To avoid branches, we might want a separate kernel launch for each boundary, similar to how we split up the loop in `apply_stencil!` (https://github.com/CliMA/ClimaCore.jl/blob/dd55a7bb2b8aa0bd5f768573c1f9c9ea8c76fd32/src/Operators/finitedifference.jl#L3337-L3345)

2. Use a single thread for each column, and have a loop over levels within the kernel. This should be the simplest (no synchronization is required), and allows dependencies.
  - I think this is what ClimateMachine.jl did
  - potentially the easiest, as we could probably just call `apply_stencil!` directly from within the kernel

Perhaps start with option 2 to get something working.

### PR DESCRIPTION
**Pull request by simonbyrne**
Fixes #1171 

- [ ] Code follows the [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) OR N/A.
- [ ] Unit tests are included OR N/A.
- [ ] Code is exercised in an integration test OR N/A.
- [ ] Documentation has been added/updated OR N/A.

### CODE REVIEW

**Comment by simonbyrne** [comment]
bors try

File: src/DataLayouts/cuda.jl [hunk 1]
```diff
@@ -92,3 +92,14 @@ function Base.copyto!(
     end
     return dest
 end
+
+function Base.copyto!(
+    dest::VF{S},
+    bc::Union{VF{S, A}, Base.Broadcast.Broadcasted{VFStyle{A}}},
+) where {S, A <: CUDA.CuArray}
+    _, _, _, Nv, Nh = size(bc)
```
Comment (sriharshakandala) [inline src/DataLayouts/cuda.jl:100]: How are we querying `Nf` here? Should `Nh` be one?

Comment (sriharshakandala) [reply src/DataLayouts/cuda.jl:100]: In general, `1D` threadblocks, with appropriate `slab` calls when direct indexing is not perimitted, may be more efficient!
```
_max_threads_cuda() = 256

function _configure_threadblock(nitems)
    nthreads = min(_max_threads_cuda(), nitems)
    nblocks = cld(nitems, nthreads)
    return (nthreads, nblocks)
end

function Base.copyto!(
    dest::VF{S},
    bc::Union{VF{S, A}, Base.Broadcast.Broadcasted{VFStyle{A}}},
) where {S, A <: CUDA.CuArray}
    nitems = length(parent(dest))
    nthreads, nblocks = _configure_thread_blocks(nitems)
    if nitems > 0
        CUDA.@cuda threads = (nthreads,) blocks = (nblocks) knl_copyto_vf!(dest, bc)
    end
    return dest
end

function knl_copyto_vf!(dest, src)
    gidx = CUDA.threadIdx().x + (CUDA.blockIdx().x - 1) * CUDA.blockDim().x
    _, _, _, Nv, _ = size(dest)
    f = cld(gidx, Nv)
    v = gidx - Nv * (f - 1)
    dest_slab = slab(dest, f)
    src_slab = slab(src, f)
    @inbounds dest_slab[v] = src_slab[v]
    return nothing
end
```

Comment (sriharshakandala) [reply src/DataLayouts/cuda.jl:100]: If operators are applies to fields (`f`s) recursively, then we may only need to distribute threads based on `Nv`!

Comment (simonbyrne) [reply src/DataLayouts/cuda.jl:100]: > How are we querying `Nf` here? Should `Nh` be one?

We don't need `Nf` (each thread can access all fields), and yes, `Nh == 1`

I agree this is inefficient (we're only doing a fraction of the work we should be doing)

Comment (simonbyrne) [reply src/DataLayouts/cuda.jl:100]: Actually, i have an idea on how to fix this.

Comment (simonbyrne) [reply src/DataLayouts/cuda.jl:100]: See #1216

**Comment by simonbyrne** [comment]
bors try

**Comment by simonbyrne** [comment]
bors try

**Comment by simonbyrne** [comment]
bors try

File: src/DataLayouts/cuda.jl (hunk 1 above)
Comment (simonbyrne) [reply src/DataLayouts/cuda.jl:100]: I realise what we have is far from optimal, but we can clean this up once we have other pieces in.

Comment (sriharshakandala) [reply src/DataLayouts/cuda.jl:100]: Sure. We can revisit this once we have other pieces in.

**Comment by simonbyrne** [comment]
bors try

**Comment by simonbyrne** [comment]
bors try

**Comment by simonbyrne** [comment]
Okay, this appears to be passing. @sriharshakandala any more comments?

**Comment by sriharshakandala** [comment]
> Okay, this appears to be passing. @sriharshakandala any more comments?

Looks good to me. The bug fix, effecting extruded 3D case, for reduction code has been added in PR https://github.com/CliMA/ClimaCore.jl/pull/1221

**Pull request merged**